lib/Email/MIME.pm

```diff
@@ -890,6 +895,21 @@ This method behaves like C<header_raw_pairs>, returning a list of field
 name/value pairs, but the values have been decoded to character strings, when
 possible.
 
+=method header_as_obj
+
+  my $first_obj = $email->header_as_obj($field);
+  my $nth_obj   = $email->header_as_obj($field, $index);
+  my @all_objs  = $email->header_as_obj($field);
+
+  my $nth_obj_of_class  = $email->header_as_obj($field, $index, $class);
+  my @all_objs_of_class = $email->header_as_obj($field, undef, $class);
+
+This method returns an object representation of the header value.  It instances
+new object via method C<from_mime_string> of specified class.  Input argument
+for that class method is list of the raw MIME-encoded values.  If class argument
+is not specified then class name is taken from the hash
+C<%Email::MIME::Header::header_to_class_map> via key field.
```

Comment:
I do not care for the third option you present.  The second is what I suggested.  The first is (effectively) sugar on the second, as the two methods must share storage, and the simplest thing is for them to be implemented in terms of the second.  They needn't be, of course.

The benefit of the first is that the registration method can detect and complain that someone is trying to change something already registered, which can prevent conflict at a distance by requiring anybody who wants to change a registration to first clear any existing one.

The benefit of the third is that one can localize the entries in the hash.

You can also localize the entries in the hash returned by a class method:

```perl
{
  my $map = Email::MIME->header_to_class_map;
  local $map->{'Message-Id'} = 'Some::Class';
  do_work;
}
```

...but this makes it clear that direct hash access is no good.  People will get the casing wrong and that will stink.  There should be a method to register and unregister, and they should case-flatten.  They should be called `set_class_for_header` and `clear_class_for_header`.

Later, we could let them be set on a per-header-object basis, maybe, if that becomes useful.